### ISSUE TITLE
Elasticsearch error with ILM enabled and dynamic indexes

### ISSUE THREAD

**Issue by irizzant** [uken/fluent-plugin-elasticsearch#830]
- [X] read [the contribution guideline](https://github.com/uken/fluent-plugin-elasticsearch/blob/master/CONTRIBUTING.md)

#### Problem

I have problems making the plugin work with ILM enabled and dynamic indexes.

I want to have fixed indexes without dates so I used [these](https://github.com/uken/fluent-plugin-elasticsearch#fixed-ilm-indices) instructions.

I used the configuration reported below and I got the following error:
```
2020-10-28 08:59:11 +0000 [error]: #0 [elasticsearch] Error while index creation - <logs.${record['kubernetes']['namespace_name']}-default-000001> error_class=Elasticsearch::Transport::Transport::Errors::Forbidden error="[403] {\"error\":{\"root_cause\":[{\"type\":\"parse_exception\",\"reason\":\"failed to parse date field [record['kubernetes']['namespace_name']] with format [uuuu.MM.dd]: [failed to parse date field [record['kubernetes']['namespace_name']] with format [uuuu.MM.dd]]\"}],\"type\":\"security_exception\",\"reason\":\"action [indices:admin/create] is unauthorized for user [elastic]\",\"caused_by\":{\"type\":\"parse_exception\",\"reason\":\"failed to parse date field [record['kubernetes']['namespace_name']] with format [uuuu.MM.dd]: [failed to parse date field [record['kubernetes']['namespace_name']] with format [uuuu.MM.dd]]\",\"caused_by\":{\"type\":\"illegal_argument_exception\",\"reason\":\"failed to parse date field [record['kubernetes']['namespace_name']] with format [uuuu.MM.dd]\",\"caused_by\":{\"type\":\"date_time_parse_exception\",\"reason\":\"Text 'record['kubernetes']['namespace_name']' could not be parsed at index 0\"}}}},\"status\":403}"
```

#### Steps to replicate

Either clone and modify https://gist.github.com/pitr/9a518e840db58f435911

**OR**

Provide example config and message

```
<match **>
      @id elasticsearch
      @type elasticsearch
      @log_level "info"
      type_name "fluentd"
      include_tag_key true
      host "elastic-cluster-es-http.elastic-system.svc.cluster.local"
      port 9200
      user "elastic"
      password xxxxxx
      scheme http
      logstash_format false
      reload_connections false
      reconnect_on_error true
      reload_on_failure true
      suppress_type_name true
      index_name "logs.${record[\'kubernetes\'][\'namespace_name\']}"
      index_date_pattern ""
      include_timestamp true
      enable_ilm true
      ilm_policy_id "fluentd-logs"
      ilm_policy_overwrite false
      log_es_400_reason true
      ilm_policy {"policy":{"phases":{"hot":{"min_age":"0ms","actions":{"rollover":{"max_age":"3d","max_size":"10gb"}}},"delete":{"min_age":"1d","actions":{"delete":{}}}}}}
      template_name "fluentd"
      template_overwrite true
      template_file "conf/template.json"
      <buffer>
        @type "file"
        path "/opt/bitnami/fluentd/logs/buffers/logs.buffer"
        flush_mode interval
        retry_type exponential_backoff
        flush_thread_count 2
        flush_interval 5s
        retry_forever 
        retry_max_interval 30
        chunk_limit_size 2M
        total_limit_size 500M
        overflow_action block
      </buffer>
    </match>
```
```
2020-10-28 08:59:11 +0000 [error]: #0 [elasticsearch] Error while index creation - <logs.${record['kubernetes']['namespace_name']}-default-000001> error_class=Elasticsearch::Transport::Transport::Errors::Forbidden error="[403] {\"error\":{\"root_cause\":[{\"type\":\"parse_exception\",\"reason\":\"failed to parse date field [record['kubernetes']['namespace_name']] with format [uuuu.MM.dd]: [failed to parse date field [record['kubernetes']['namespace_name']] with format [uuuu.MM.dd]]\"}],\"type\":\"security_exception\",\"reason\":\"action [indices:admin/create] is unauthorized for user [elastic]\",\"caused_by\":{\"type\":\"parse_exception\",\"reason\":\"failed to parse date field [record['kubernetes']['namespace_name']] with format [uuuu.MM.dd]: [failed to parse date field [record['kubernetes']['namespace_name']] with format [uuuu.MM.dd]]\",\"caused_by\":{\"type\":\"illegal_argument_exception\",\"reason\":\"failed to parse date field [record['kubernetes']['namespace_name']] with format [uuuu.MM.dd]\",\"caused_by\":{\"type\":\"date_time_parse_exception\",\"reason\":\"Text 'record['kubernetes']['namespace_name']' could not be parsed at index 0\"}}}},\"status\":403}"
```

#### Expected Behavior or What you need to ask

Fluentd should be able to create the index in Elasticsearch

#### Using Fluentd and ES plugin versions

* OS version
* Bare Metal or within Docker or Kubernetes or others?
kubernetes v1.18.9
* Fluentd v0.12 or v0.14/v1.0
  * paste result of ``fluentd --version`` or ``td-agent --version``
fluentd 1.11.4
* ES plugin 3.x.y/2.x.y or 1.x.y
  * paste boot log of fluentd or td-agent
  * paste result of ``fluent-gem list``, ``td-agent-gem list`` or your Gemfile.lock
elasticsearch (7.9.0)
elasticsearch-api (7.9.0)
elasticsearch-transport (7.9.0)
elasticsearch-xpack (7.9.0)
fluent-plugin-elasticsearch (4.2.2)

* ES version (optional) 7.9.0
* ES template(s) (optional)

**Comment by cosmo0920** [comment]
> [failed to parse date field [record['kubernetes']['namespace_name']] with format [uuuu.MM.dd]]\"}]

ES cluster complains that received date format should be invalid.

**Comment by irizzant** [comment]
I'm using `index_name` like this:
`index_name "logs.${record[\'kubernetes\'][\'namespace_name\']}"`

it should take the value of kubernetes namespace and put it in the index name, how come this is being used as a timestamp format?

**Comment by cosmo0920** [comment]
> it should take the value of kubernetes namespace and put it in the index name, how come this is being used as a timestamp format?

I guess that ES cluster might misinterpret as date format patterns.

Umm..., how about using the following configuration?


```aconf
<match **>
      @id elasticsearch
      @type elasticsearch
      @log_level "info"
      type_name "fluentd"
      include_tag_key true
      host "elastic-cluster-es-http.elastic-system.svc.cluster.local"
      port 9200
      user "elastic"
      password xxxxxx
      scheme http
      logstash_format false
      reload_connections false
      reconnect_on_error true
      reload_on_failure true
      suppress_type_name true
      index_name logs.${$.kubernetes.namespace_name}
      index_date_pattern ""
      include_timestamp true
      enable_ilm true
      ilm_policy_id "fluentd-logs"
      ilm_policy_overwrite false
      log_es_400_reason true
      ilm_policy {"policy":{"phases":{"hot":{"min_age":"0ms","actions":{"rollover":{"max_age":"3d","max_size":"10gb"}}},"delete":{"min_age":"1d","actions":{"delete":{}}}}}}
      template_name "fluentd"
      template_overwrite true
      template_file "conf/template.json"
      <buffer tag, $.kubernetes.namespace_name>
        @type "file"
        path "/opt/bitnami/fluentd/logs/buffers/logs.buffer"
        flush_mode interval
        retry_type exponential_backoff
        flush_thread_count 2
        flush_interval 5s
        retry_forever 
        retry_max_interval 30
        chunk_limit_size 2M
        total_limit_size 500M
        overflow_action block
      </buffer>
    </match>
```

Using nested accessors in built-in placeholder, you should include nested records information in buffer metadata.
These metadata should be included with buffer section attributes: `tag, $.kubernetes.namespace_name`.

see also: https://docs.fluentd.org/configuration/buffer-section#nested-field-support

**Comment by irizzant** [comment]
With that configuration I still get:
```
2020-11-02 13:41:29.686301714 +0000 fluent.warn: {"retry_time":4,"next_retry_seconds":"2020-11-02 13:41:37 +0000","chunk":"5b2ff230d05ea6dd1c7a77414e856b77","error":"#<Elasticsearch::Transport::Transport::Errors::Forbidden: [403] {\"error\":{\"root_cause\":[{\"type\":\"parse_exception\",\"reason\":\"failed to parse date field [$.kubernetes.namespace_name] with format [uuuu.MM.dd]: [failed to parse date field [$.kubernetes.namespace_name] with format [uuuu.MM.dd]]\"}],\"type\":\"security_exception\",\"reason\":\"action [indices:admin/aliases] is unauthorized for user [elastic]\",\"caused_by\":{\"type\":\"parse_exception\",\"reason\":\"failed to parse date field [$.kubernetes.namespace_name] with format [uuuu.MM.dd]: [failed to parse date field [$.kubernetes.namespace_name] with format [uuuu.MM.dd]]\",\"caused_by\":{\"type\":\"illegal_argument_exception\",\"reason\":\"failed to parse date field [$.kubernetes.namespace_name] with format [uuuu.MM.dd]\",\"caused_by\":{\"type\":\"date_time_parse_exception\",\"reason\":\"Text '$.kubernetes.namespace_name' could not be parsed at index 0\"}}}},\"status\":403}>","message":"[elasticsearch] failed to flush the buffer. retry_time=4 next_retry_seconds=2020-11-02 13:41:37 +0000 chunk=\"5b2ff230d05ea6dd1c7a77414e856b77\" error_class=Elasticsearch::Transport::Transport::Errors::Forbidden error=\"[403] {\\\"error\\\":{\\\"root_cause\\\":[{\\\"type\\\":\\\"parse_exception\\\",\\\"reason\\\":\\\"failed to parse date field [$.kubernetes.namespace_name] with format [uuuu.MM.dd]: [failed to parse date field [$.kubernetes.namespace_name] with format [uuuu.MM.dd]]\\\"}],\\\"type\\\":\\\"security_exception\\\",\\\"reason\\\":\\\"action [indices:admin/aliases] is unauthorized for user [elastic]\\\",\\\"caused_by\\\":{\\\"type\\\":\\\"parse_exception\\\",\\\"reason\\\":\\\"failed to parse date field [$.kubernetes.namespace_name] with format [uuuu.MM.dd]: [failed to parse date field [$.kubernetes.namespace_name] with format [uuuu.MM.dd]]\\\",\\\"caused_by\\\":{\\\"type\\\":\\\"illegal_argument_exception\\\",\\\"reason\\\":\\\"failed to parse date field [$.kubernetes.namespace_name] with format [uuuu.MM.dd]\\\",\\\"caused_by\\\":{\\\"type\\\":\\\"date_time_parse_exception\\\",\\\"reason\\\":\\\"Text '$.kubernetes.namespace_name' could not be parsed at index 0\\\"}}}},\\\"status\\\":403}\""}
```

**Comment by irizzant** [comment]
I'm still trying but I only managed to get close to it using the $tag variable as index name, something like this:
```
<match **>
    @id elasticsearch
    @type elasticsearch
    @log_level info
    include_tag_key true
    host elasticsearch.elasticsearch
    port 9200
    default_elasticsearch_version 7
    verify_es_version_at_startup true
    suppress_type_name true
    logstash_format true
    logstash_prefix logs.${tag}
    logstash_prefix_separator .
    enable_ilm true
    ilm_policy_id fluentd-logs
    ilm_policy {"policy":{"phases":{"hot":{"min_age":"0ms","actions":{"rollover":{"max_age":"3d","max_size":"20gb"},"set_priority":{"priority":100}}},"warm":{"actions":{"allocate":{"include":{},"exclude":{},"require":{"data":"warm"}},"set_priority":{"priority":50}}},"delete":{"min_age":"30d","actions":{"delete":{}}}}}}
    ilm_policy_overwrite false
    log_es_400_reason true
    # Disabled until https://github.com/uken/fluent-plugin-elasticsearch/issues/798 is fixed
    # templates { "logs": "/etc/fluent/config.d/logs-es-template.json", "formative-logs": "/etc/fluent/config.d/formative-es-template.json", "webclient-logs": "/etc/fluent/config.d/webclient-es-template.json" }
    template_overwrite true
    template_name logs
    template_file "/etc/fluent/config.d/logs-es-template.json"

    <buffer tag>
      @type file
      path /var/log/fluentd-buffers/kubernetes.${tag}
      flush_mode interval
      retry_max_interval 30
      flush_thread_count 2
    </buffer>
  </match>
```

The problem with this configuration is that, at least in my case, the number of created shards exponentially grows until it hits the default 3000 limit.

**Comment by metalrex100** [comment]
@irizzant for using vars in elasticsearch output you should change `@type` from `elasticsearch` to `elasticsearch_dynamic`

**Comment by cosmo0920** [comment]
@metalrex100  Please don't suggest outdated plugin. Thanks.

**Comment by metalrex100** [comment]
@metalrex100 does elasticsearch plugin currently support dynamic values?

**Comment by metalrex100** [comment]
@irizzant have you set up index template in elastic search?
You can try to setup index template and ILM policy manually with elasticsearch API.

I'm currently setting up fluentd (fluentd daemonset in Kibernetes) with ELK.
I manually created index template and ILM policy. And everything is working for me.

**Comment by metalrex100** [comment]
@cosmo0920 then you should also add `deprecated` label to this section in docs?
https://github.com/uken/fluent-plugin-elasticsearch/tree/v4.2.2#dynamic-configuration

**Comment by cosmo0920** [comment]
Thanks for the feedback. I'd added a note for to be going to be marked as deprecated. https://github.com/uken/fluent-plugin-elasticsearch/commit/6a20812bb390dd80ca3356ed76ece9c01c9de8a2

**Comment by irizzant** [comment]
Hello,
so still no way to configure this scenario?

**Comment by metalrex100** [comment]
@irizzant if I'm right, you want to configure fluentd-elasticsearch plugin to create separate indexes for each kubernetes cluster and the index name should have name `logs.cluster_name`.

I'm currently testing it and will give feedback about results!

**Comment by irizzant** [comment]
@metalrex100 thanks, but in reality I'm trying to create separate indexes for each kubernetes namespace / pod, the indexes should be named like `logs.namespace` or `logs.namespace.pod`.

Thanks

**Comment by metalrex100** [comment]
@irizzant yeap, sorry.) I have more or less similar goal: I want to create index for each microservice. So in my case the index name should be `fluentd-appname`.

As far as I understood it is impossible to make multiple rollovered indexes under same index template.

**Explanation why:**
To be rollovered index must have same **alias** as **rollover_alias** setting of it's template.
To init rollover you have to create initial write index with `-000001` postfix, so ILM could increment it on each rollover.

For example you have:
- ILM policy named `fluentd-namespaces`
- index template with following lifecycle setitngs:
```json
"lifecycle": {
    "name": "fluentd-namespaces",
    "rollover_alias": "fluentd-namespace"
}
```

You want to create multiple indexes with names like `fluentd-namespaceA`, `fluentd-namespaceB`, `fluentd-namespaceC`,  and rollover them.

So the result you would like to see in your indexes-list is:
```
fluentd-namespaceA-000001
fluentd-namespaceA-000002
fluentd-namespaceA-000003
fluentd-namespaceB-000001
fluentd-namespaceB-000002
fluentd-namespaceB-000003
fluentd-namespaceC-000001
fluentd-namespaceC-000002
fluentd-namespaceC-000003
```

The problem is that each index (`fluentd-namespaceA-000001`, `fluentd-namespaceB-000001`, `fluentd-namespaceC-000001`) to be rollovered must have same **alias** as **rollover_alias** of the index template.
If all indexes have same alias, then **plugin will create only first index** (`fluentd-namespaceA-000001`). Other indexes (`fluentd-namespaceB-000001`, `fluentd-namespaceC-000001`) will be ignored because alias they want to use is already created (for `fluentd-namespaceA-000001` index)

And it is actually correct behavior of the plugin, because you can't add document to index by alias if alias belongs to several indexes. It's obvious because Elasticsearch just does not know to which index document should be added.

More about **one alias with multiple indexes**
https://discuss.elastic.co/t/one-alias-with-multiple-indices/22230

Hope my explanation is more or less understandable)

**Comment by metalrex100** [comment]
So as I can see there are two solutions:
- keep all data in singe index (fluentd-allnamespaces) with single index template.
- create separate index template for each namespace. Each index template will have its unique `rollover_alias` and each namespace index will also have its alias. But then you have to create template file for each index template to be able to specify correct index pattern. If all index templates will have same pattern (e.g. `fluentd-namespace*`) - they will collide between each other.

Actually there is third solutions - **data streams**. They handle index creation and rollover by themselves.
More about data streams:
https://www.elastic.co/guide/en/elasticsearch/reference/current/data-streams.html

But currently this plugin **does not support data streams**.

**Comment by irizzant** [comment]
Thanks @metalrex100 .

The explanation makes sense, and as far as I can understand you didn't actually get to a solution for your scenario either.

Data streams rather belongs to time-series world, and manually creating data for each namespace does require extra work.

I think I'm going to go down the single index path, frankly it works just fine and if you keep the index small enough there shouldn't be any performance impact after all.

**Comment by metalrex100** [comment]
@irizzant I guess task may be closed?
@cosmo0920 cc

**Comment by irizzant** [comment]
@metalrex100 about your explanation above, there is something I don't understand.
I have updated elastic to 7.10, and it looks like the ILM stopped working with the same config, which is weird.
In particular, fluentd is failing with:
```
2020-12-04 14:29:33 +0000 [error]: #0 unexpected error error_class=Elasticsearch::Transport::Transport::Errors::BadRequest error="[400] {\"error\":{\"root_cause\":[{\"type\":\"invalid_alias_name_exception\",\"reason\":\"Invalid alias name [fluentd], an index exists with the same name as the alias\",\"index_uuid\":\"0MjBXrYQTXu8q76iQE0cYQ\",\"index\":\"fluentd\"}],\"type\":\"invalid_alias_name_exception\",\"reason\":\"Invalid alias name [fluentd], an index exists with the same name as the alias\",\"index_uuid\":\"0MjBXrYQTXu8q76iQE0cYQ\",\"index\":\"fluentd\"},\"status\":400}"
```
It's like the plugin is trying to create the index (which exists in elastic with name fluentd)

**Comment by metalrex100** [comment]
@irizzant according to the error message it seems like your index name is equal to the alias name.

It's weird, because plugin should create index with `-000001` postfix and then index name and alias name will not be the same.
But such situation may occur if you removed index manually in elasticsearch and did not restart fluentd daemonset. (or container).
Because this plugin caches created indexes. It means that if rollover index (with `-000001` postfix) has been created by plugin, then it will not be created if you will manually removed your indexes from elasticsearch. Instead plugin will try to push message via alias (which is taken from you `index_name` param). And since no indexes exist, elasticsearch will try to create index with name which will equal to already existing alias.

**Comment by irizzant** [comment]
@metalrex100 it's definitively weird.

I upgraded elastic to 7.10 and I didn't restart the fluentd pods as you said.
There is an index in my elastic which doesn't have any ILM policy associated to it and it's called **fluentd**.
So this is what is causing the name clash with the alias the plugin is trying to create.
Conversely, there is also an index **fluentd-000001** which has got a `logstash_policy` ILM policy associated, but that policy does not exist!
I removed the fluentd index and fluentd now starts correctly, but it didn't recreate the ILM policy `logstash_policy`.
I tried to restart fluentd pods but it didn't recreate the ILM policy.
To make sure that index was created the right way at least, I removed also the `fluentd-000001` and waited for it to be recreated.
Fluentd created `fluentd` index instead!
I had to remove everything:
* index template
* index
* ILM policy
and after I restarted fluentd pod I finally got the `fluentd-000001` index created with the right ILM policy associated to it, but my logs were lost.

**Comment by metalrex100** [comment]
@irizzant any updates should be done firstly in test env)
Glad you didn’t lose important data. If you want to report this problem - I believe new issue should be created and this one should be closed.

**Comment by irizzant** [comment]
> any updates should be done firstly in test env
@metalrex100 I totally agree!

Regarding this issue I'm going to close it. I don't know if I'll be able to replicate the problem described above for the new issue though.

**Comment by aalemanq** [comment]
Hello, I have similar problem when I try to create templates with $tag.

I let the info here but..is closed :(

https://github.com/uken/fluent-plugin-elasticsearch/issues/866
